// eslint-plugin-lit/lib/rules/no-classfield-shadowing.js
"use strict";
/**
 * @fileoverview Disallows properties shadowed as class fields
 * @author Michel Langeveld <https://github.com/michellangeveld>
 */
const util_1 = require("../util");
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
    meta: {
        docs: {
            description: 'Disallows properties shadowed as class fields',
            recommended: true,
            url: 'https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/no-classfield-shadowing.md'
        },
        schema: [],
        messages: {
            noClassfieldShadowing: 'The {{ prop }} property is a class field which has the same name as ' +
                'static property which could have unintended side-effects.'
        }
    },
    create(context) {
        return {
            ClassDeclaration: (node) => {
                if ((0, util_1.isLitClass)(node)) {
                    const propertyMap = (0, util_1.getPropertyMap)(node);
                    const classMembers = (0, util_1.getClassFields)(node);
                    for (const [prop, { key }] of propertyMap.entries()) {
                        if (classMembers.has(prop)) {
                            context.report({
                                node: key,
                                messageId: 'noClassfieldShadowing',
                                data: { prop }
                            });
                        }
                    }
                }
            }
        };
    }
};
module.exports = rule;
