// fprint/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
const path = require("path");
const fs_1 = require("fs");
const callsite = require("callsite");
const isstream = require("isstream");
/**
 * Utility function to create stream based on file's location, supports both relative and absolute paths
 * @param  {String} location [description]
 * @return {ReadStream}
 */
function createStream(location) {
    if (path.isAbsolute(location)) {
        return fs_1.createReadStream(location);
    }
    // For relative paths
    const stack = callsite();
    const length = stack.length;
    // Filter out the file itself
    let iterator = 0;
    while (iterator < length) {
        const call = stack[iterator++];
        const filename = call.getFileName();
        if (__filename !== filename) {
            const source = path.resolve(path.dirname(filename), location);
            return fs_1.createReadStream(source);
        }
    }
}
/**
 * Accepts Buffer and creates it's digest
 * @param  {Buffer}   buffer
 * @param  {string}   algorithm
 * @return {string}
 */
function digestSync(buffer, algorithm) {
    const hash = crypto_1.createHash(algorithm).update(buffer);
    return hash.digest('hex');
}
exports.digestSync = digestSync;
/**
 * Accepts stream and creates it's digest
 * @param  {ReadStream}   stream
 * @param  {string}           algorithm
 * @return {Promise<string>}
 */
function digestStream(stream, algorithm) {
    return new Promise((resolve, reject) => {
        const hash = crypto_1.createHash(algorithm);
        stream.on('data', (chunk) => hash.update(chunk));
        stream.on('error', (error) => reject(error));
        stream.on('end', () => resolve(hash.digest('hex')));
    });
}
exports.digestStream = digestStream;
function isReadable(object) {
    return isstream.isReadable(object);
}
function isString(x) {
    return typeof x === 'string';
}
/**
 * Creates fingerprint for a given input
 * @param  {Buffer|string|ReadableStream}   source
 * @param  {string}                         algorithm - any supported algorithm by node.js crypto module https://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm
 * @return {Promise<string>}
 */
async function createFingerprint(source, algorithm) {
    // We have file buffered in memory, create digest on it
    if (Buffer.isBuffer(source)) {
        return digestSync(source, algorithm);
    }
    if (isReadable(source)) {
        return digestStream(source, algorithm);
    }
    if (isString(source)) {
        const stream = createStream(source);
        if (stream) {
            return digestStream(stream, algorithm);
        }
    }
    throw new TypeError('source must be either a buffer, readable stream or a string with absolute or relative path');
}
exports.createFingerprint = createFingerprint;
exports.default = createFingerprint;
